handlers/user.go

```diff
@@ -1414,67 +1422,96 @@ func UserConfirmUpdateEmail(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	hash := vars["hash"]
 
-	_, session, err := getUserSession(r)
+	sessionUser, session, err := getUserSession(r)
 	if err != nil {
-		logger.Errorf("error retrieving session: %v", err)
+		utils.LogError(err, "error retrieving session for email update confirmation", 0, map[string]interface{}{"hash": hash})
 		http.Error(w, "Internal server error", http.StatusInternalServerError)
 		return
 	}
+	if !sessionUser.Authenticated {
+		utils.SetFlash(w, r, authSessionName, "Error: You need to be logged in to update your email. Please login and try again.")
+		http.Redirect(w, r, "/login", http.StatusSeeOther)
+		return
+	}
 
 	user := struct {
-		ID        int64     `db:"id"`
-		Email     string    `db:"email"`
-		ConfirmTs time.Time `db:"email_confirmation_ts"`
-		Confirmed bool      `db:"email_confirmed"`
-		NewEmail  string    `db:"email_change_to_value"`
+		ID               int64     `db:"id"`
+		Email            string    `db:"email"`
+		ConfirmTs        time.Time `db:"email_confirmation_ts"`
+		Confirmed        bool      `db:"email_confirmed"`
+		NewEmail         string    `db:"email_change_to_value"`
+		StripeCustomerId string    `db:"stripe_customer_id"`
 	}{}
 
-	err = db.FrontendWriterDB.Get(&user, "SELECT id, email, email_confirmation_ts, email_confirmed, email_change_to_value FROM users WHERE email_confirmation_hash = $1", hash)
+	err = db.FrontendWriterDB.Get(&user, "SELECT id, email, COALESCE(email_confirmation_ts, '399-01-01 BC'::timestamp), email_confirmed, COALESCE(email_change_to_value, ''), COALESCE(stripe_customer_id, '') FROM users WHERE email_confirmation_hash = $1", hash)
 	if err != nil {
-		logger.Errorf("error retrieving email for confirmation_hash %v %v", hash, err)
-		utils.SetFlash(w, r, authSessionName, "Error: This confirmation link is invalid / outdated.")
-		http.Redirect(w, r, "/confirmation", http.StatusSeeOther)
+		if err != sql.ErrNoRows {
+			utils.LogError(err, "error retrieving user data for updating email", 0, map[string]interface{}{"hash": hash, "userID": sessionUser.UserID})
+		}
+		utils.SetFlash(w, r, authSessionName, "Error: This link is invalid / outdated.")
+		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
+		return
+	}
+
+	// validate data
+
+	if sessionUser.UserID != uint64(user.ID) {
+		utils.SetFlash(w, r, authSessionName, "Error: This link is invalid / outdated.")
```

Comment:
**minor:** Maybe a confusing flash message since as a user I would not know that the reason for the failure was that a different user was logged in.

Reply:
This has security implications. If an authenticated user would get another message they would know that this is the valid hash for some other user, which we probably don't want.

Reply:
Thank you for the clarification.